### PR TITLE
mat64: add support for vector shadowing detection

### PR DESCRIPTION
**Pull request by kortschak**
Updates #336.

@btracey @vladimir-ch Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
I am reasonably sure I got all the Vector/Dense candidates, but please take extra looking for cases I've missed.

File: mat64/lq.go [hunk 1]
```diff
@@ -210,7 +213,7 @@ func (v *Vector) SolveLQVec(lq *LQ, trans bool, b *Vector) error {
 	} else {
 		v.reuseAs(c)
 	}
-	m := vecAsDense(v)
-	bm := vecAsDense(b)
+	m := v.asDense()
```
Comment (kortschak) [inline mat64/lq.go]: Considering moving these into the call line as `v.asDense().SolveLQ(lq, trans, b.asDense())`.

File: mat64/qr.go [hunk 2]
```diff
@@ -209,21 +212,7 @@ func (v *Vector) SolveQRVec(qr *QR, trans bool, b *Vector) error {
 	} else {
 		v.reuseAs(c)
 	}
-	m := vecAsDense(v)
-	bm := vecAsDense(b)
+	m := v.asDense()
+	bm := b.asDense()
```
Comment (kortschak) [inline mat64/qr.go]: Considering moving these into the call line as `return v.asDense().SolveQR(qr, trans, b.asDense())`.

File: mat64/solve.go [hunk 3]
```diff
@@ -107,17 +107,20 @@ func (m *Dense) Solve(a, b Matrix) error {
...
 	v.reuseAs(c)
-	m := vecAsDense(v)
+	m := v.asDense()
 	// We conditionally create bm as m when b and v are identical
 	// to prevent the overlap detection code from identifying m
 	// and bm as overlapping but not identical.
 	bm := m
 	if v != b {
-		bm = vecAsDense(b)
+		bm = b.asDense()
```
Comment (kortschak) [inline mat64/solve.go]: Not here because of conditionality.

**Comment by vladimir-ch** [comment]
Not sure: check overlap with `lu.lu` in `LU.RankOne`?

**Comment by kortschak** [comment]
I don't think it's possible for there to be an overlap there because the caller can't modify that or recover it. @btracey?

**Comment by vladimir-ch** [comment]
Oh, sorry for the noise, it makes a copy of x and y into xs and ys.

**Comment by vladimir-ch** [comment]
Your reason being the relevant one, of course.

**Comment by vladimir-ch** [comment]
What about in `Dense.Mul`? There are checks when a and b are RawMatrixer but not when RawVectorer or *Vector.

**Comment by kortschak** [comment]
Thanks.

**Comment by kortschak** [comment]
Related to that, we don't actually use RawVectorer much - the concrete type is used in all cases. Why?

**Comment by vladimir-ch** [comment]
I have wondered that myself before, I don't know. Maybe history, maybe because RawVector doesn't carry the length. In Mul it is not needed but at several places in matrix.go the `n` field is used (could be avoided though). I think we should use RawVectorer instead.

I don't see any other missed overlap checks and I'm fine with the one-liners. LGTM but needless to say, wait for @btracey .

**Comment by kortschak** [comment]
I think my question was more about why RawVectorer exists.

**Comment by btracey** [comment]
I agree that our position with Vector is weird. However, why shouldn't RawVectorer exist when all all of the other RawX types do?

**Comment by btracey** [comment]
I agree that lu shouldn't need it since it is impossible for the user to access the data directly.

**Comment by btracey** [comment]
LGTM

**Pull request merged**